lib/web_ui/lib/src/engine/bitmap_canvas.dart

```diff
@@ -664,6 +664,201 @@ class BitmapCanvas extends EngineCanvas with SaveStackTracking {
+          ? vertices.positions
+          : _convertVertexPositions(mode, vertices.positions);
+      // Draw hairline for vertices if no vertex colors are specified.
+      _drawHairline(positions, paint.color ?? ui.Color(0xFF000000));
+      return;
+    }
+
+    final html.CanvasElement glCanvas = html.CanvasElement(
+      width: _widthInBitmapPixels,
+      height: _heightInBitmapPixels,
+    );
+
+    glCanvas.style
+      ..position = 'absolute'
+      ..width = _canvas.style.width
+      ..height = _canvas.style.height;
+    glCanvas.className = 'gl-canvas';
+
+    _children.add(glCanvas);
+    rootElement.append(glCanvas);
+
+    _GlContext gl = _GlContext(glCanvas);
+
+    // Create and compile shaders.
+    Object vertexShader =
+        gl.compileShader('VERTEX_SHADER', _vertexShaderTriangle);
+    Object fragmentShader =
+        gl.compileShader('FRAGMENT_SHADER', _fragmentShaderTriangle);
+    // Create a gl program and link shaders.
+    Object program = gl.createProgram();
+    gl.attachShader(program, vertexShader);
+    gl.attachShader(program, fragmentShader);
+    gl.linkProgram(program);
+    gl.useProgram(program);
+
+    // Set uniform to scale 0..width/height pixels coordinates to -1..1
+    // clipspace range and flip the Y axis.
+    Object resolution = gl.getUniformLocation(program, 'u_scale');
+    gl.setUniform4f(resolution, 2.0 / _widthInBitmapPixels.toDouble(),
+        -2.0 / _heightInBitmapPixels.toDouble(), 1, 1);
+    Object shift = gl.getUniformLocation(program, 'u_shift');
+    gl.setUniform4f(shift, -1, 1, 0, 0);
+
+    // Setup geometry.
+    Object positionsBuffer = gl.createBuffer();
+    assert(positionsBuffer != null);
+    gl.bindArrayBuffer(positionsBuffer);
+    final Float32List positions = vertices.positions;
+    gl.bufferData(positions, gl.kStaticDraw);
+    js_util.callMethod(
+        gl.glContext, 'vertexAttribPointer', [0, 2, gl.kFloat, false, 0, 0]);
+    gl.enableVertexAttribArray(0);
+
+    // Setup color buffer.
+    Object colorsBuffer = gl.createBuffer();
+    gl.bindArrayBuffer(colorsBuffer);
+    // Buffer kBGRA_8888.
+    gl.bufferData(colors, gl.kStaticDraw);
+
+    js_util.callMethod(gl.glContext, 'vertexAttribPointer',
+        [1, 4, gl.kUnsignedByte, true, 0, 0]);
+    gl.enableVertexAttribArray(1);
+    gl.clear();
+    final int vertexCount = positions.length ~/ 2;
+    gl.drawTriangles(vertexCount, mode);
+  }
+
+  void _drawHairline(Float32List positions, ui.Color color) {
+    assert(positions != null);
+    html.CanvasRenderingContext2D _ctx = ctx;
+    save();
+    final int pointCount = positions.length ~/ 2;
+    _setFillAndStrokeStyle('', color.toCssString());
+    _ctx.lineWidth = 1.0;
+    int triIndex = 0;
+    _ctx.beginPath();
+    for (int i = 0, len = pointCount * 2; i < len; i += 2) {
+      final double dx = positions[i];
+      final double dy = positions[i + 1];
+      if (triIndex & 3 == 0) {
```

Comment:
is `& 3` necessary? It seems like `triIndex` will always be between 0 and 2

Reply:
% 2 , & 3 same here. Switched to 2 loops to make it more readable.